File: paack-java-sdk/src/main/java/eu/paack/sdk/api/CoverageApi.java
package eu.paack.sdk.api;

import com.fasterxml.jackson.core.type.TypeReference;
import eu.paack.sdk.PaackConstants;
import eu.paack.sdk.api.model.response.CheckCoverageResponse;
import eu.paack.sdk.api.model.response.Error;
import eu.paack.sdk.api.model.response.PaackResponse;
import eu.paack.sdk.exceptions.ApiException;
import eu.paack.sdk.model.PaackCoverage;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import lombok.extern.slf4j.Slf4j;
import org.apache.http.NameValuePair;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 * Verifies whether the postcode or postcode zone specified in the query parameter is covered by Paack.
 * Returns all the covered postcodes and postcode zones if no query parameter is specified.
 *
 * More details can be found in the Paack API documentation: https://paack.readme.io/reference/coverage
 */
@Slf4j
@SuperBuilder
@NoArgsConstructor
public class CoverageApi extends PaackApi {

    /**
     * Returns all covered postcodes and postcode zones.
     *
     * @return
     */
    public PaackResponse<CheckCoverageResponse, Error> checkCoverage() {
        try {
            PaackResponse<CheckCoverageResponse, Error> response = apiClient.invokeAPI(PaackEndpoint.coverage,
                    "GET",
                    null,
                    null,
                    null,
                    new TypeReference<CheckCoverageResponse>() {
                    });
            log.info(response.toString());
            return PaackResponse.<CheckCoverageResponse, Error>builder()
                    .data(response.getData())
                    .build();
        } catch (ApiException e) {
            log.error("Coverage service failed to load", e);
            return errorMessage("CoverageApi.checkCoverage", "Failed to ge the data:" + e.getMessage());
        }
    }

    /**
     * For a given coverage_zone we need then to check its coverage not only by the complete coverage_zone.
     *
     * Returns True for coverage or False for no coverage.
     *
     * Args:
     * country (str): Country code in ISO 3166 alpha 2 format
     * 	coverage_zone (str):  Coverage zone
     * @param country
     * @param coverage_zone
     * @return
     */
    public PaackResponse<Boolean, Error> checkCoverageZone(String country, String coverage_zone) {
        if (country == null || country.length() == 0) {
            return errorMessage("Country cannot be empty", "CoverageAPI.checkCoverageZone.country", "001");
        }

        if (coverage_zone == null || coverage_zone.length() == 0) {
            return errorMessage("Coverage zone cannot be empty", "CoverageAPI.checkCoverageZone.country", "001");
        }
        try {
            PaackResponse<PaackCoverage, Error> response = apiClient.invokeAPI(PaackEndpoint.coverage,
                    "GET",
                    null,
                    mapQueryParamsZone(country, coverage_zone),
                    null,
                    new TypeReference<PaackCoverage>() {
                    });
            log.info(response.toString());

            if (response.getData() != null) {
                return PaackResponse.<Boolean, Error>builder()
                        .data("coverage".equalsIgnoreCase(response.getData().getMessage()))
                        .build();
            } else
                return errorMessage("CoverageApi.getCoveragePostalCode", "Failed to ge the data:" + response.getError().toString());
        } catch (ApiException e) {
            log.error("Coverage service failed to load", e);
            return errorMessage("CoverageApi.checkCoverageZone", "Failed to ge the data:" + e.getMessage());
        }
    }

    /**
     * For a given postcode we need then to check its coverage not only by the complete postcode.
     *
     * Returns True for coverage or False for no coverage.
     *
     * Args:
     * country (str): Country code in ISO 3166 alpha 2 format
     * coverage_code (str): Postcode of the location
     * @param country
     * @param coverage_code
     * @return
     */
    public PaackResponse<Boolean, Error> getCoveragePostalCode(String country, String coverage_code) {
        if (country == null || country.length() == 0) {
            return errorMessage("Country cannot be empty", "CoverageAPI.getCoveragePostalCode.country", "001");
        }

        if (coverage_code == null || coverage_code.length() == 0) {
            return errorMessage("Coverage code cannot be empty", "CoverageAPI.getCoveragePostalCode.country", "001");
        }
        try {
            PaackResponse<PaackCoverage, Error> response = apiClient.invokeAPI(PaackEndpoint.coverage,
                    "GET",
                    null,
                    mapQueryParamsCode(country, coverage_code),
                    null,
                    new TypeReference<PaackCoverage>() {
                    });
            log.info(response.toString());
            if (response.getData() != null) {
                return PaackResponse.<Boolean, Error>builder()
                        .data("coverage".equalsIgnoreCase(response.getData().getMessage()))
                        .build();
            } else
                return errorMessage("CoverageApi.getCoveragePostalCode", "Failed to ge the data:" + response.getError().toString());

        } catch (ApiException e) {
            log.error("Coverage service failed to load", e);
            return errorMessage("CoverageApi.getCoveragePostalCode", "Failed to ge the data:" + e.getMessage());
        }
    }

    private List<NameValuePair> mapQueryParamsCode(String country, String coverage_code) {
        return Stream.of(
                param(PaackConstants.PARAM_COUNTRY, country),
                param(PaackConstants.PARAM_COVERAGE_CODE, coverage_code)
        ).collect(Collectors.toList());
    }

    private List<NameValuePair> mapQueryParamsZone(String country, String coverage_zone) {
        return Stream.of(
                param(PaackConstants.PARAM_COUNTRY, country),
                param(PaackConstants.PARAM_COVERAGE_ZONE, coverage_zone)
        ).collect(Collectors.toList());
    }

}

File: paack-go-sdk/rest/coverage_api.go
package rest

import (
	"encoding/json"
	"fmt"
	"github.com/PaackEng/paack-go-sdk/api_client"
	"github.com/PaackEng/paack-go-sdk/logger"
	"github.com/PaackEng/paack-go-sdk/mixin"
	"github.com/PaackEng/paack-go-sdk/types/responses"
	"io"
	"strconv"
	"strings"
)

/*
CoverageApi - Verifies whether the postcode or postcode zone specified in the query parameter is covered by Paack.
Returns all the covered postcodes and postcode zones if no query parameter is specified.

More details can be found in the Paack API documentation: https://paack.readme.io/reference/coverage
*/
type CoverageApi struct {
	BaseApi
}

func NewCoverageApi(audience, domain string, path *mixin.Resources, client api_client.ApiClient, logger logger.Logger) *CoverageApi {

	return &CoverageApi{
		*NewBaseApi(client, domain, audience, path, logger),
	}
}

/*
CheckCoverage - Returns all covered postcodes and postcode zones.

Returns:

	response (CheckCoverageResponse | ErrorResponse): CoverageResponse({:coverage_codes, :coverage_zones})
*/
func (t *CoverageApi) CheckCoverage() (successResponse *responses.CheckCoverageResponse, errorResponse *responses.ErrorResponse) {
	t.logger.Info("Calling CoverageApi.CheckCoverage method")

	url := fmt.Sprintf("%s%s", t.domain, t.path.Coverage)

	resp, err := t.client.Get(url, true, t.audience)
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", t.logger)
		return
	}

	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", body), strconv.Itoa(resp.StatusCode), "", t.logger)
		return
	}

	err = json.Unmarshal(body, &successResponse)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", t.logger)
		return
	}

	return
}

/*
CheckCoveragePostalCode - For a given postcode we need then to check its coverage not only by the complete postcode

Returns:

	True for coverage or False for no coverage.

Args:

	country (str): Country code in ISO 3166 alpha 2 format
	coverage_code (str): Postcode of the location
*/
func (t *CoverageApi) CheckCoveragePostalCode(country, coverageCode string) (successResponse bool, errorResponse *responses.ErrorResponse) {
	t.logger.Info("Calling CoverageApi.CheckCoveragePostalCode method")

	if len(country) == 0 {
		errorResponse = responses.NewError("country cannot be empty", "001", "country", t.logger)
		return
	}

	if len(coverageCode) == 0 {
		errorResponse = responses.NewError("coverageCode cannot be empty", "001", "coverageCode", t.logger)
		return
	}

	country = t.checkCountry(country)
	url := fmt.Sprintf("%s%s?country=%s&coverage_code=%s", t.domain, t.path.Coverage, country, coverageCode)

	resp, err := t.client.Get(url, true, t.audience)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		if resp.StatusCode == 404 {
			return
		} else {
			errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", resp.Status), strconv.Itoa(resp.StatusCode), "", t.logger)
			return
		}
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", t.logger)
		return
	}

	var response *responses.CheckCoveragePostalCodeResponse
	err = json.Unmarshal(body, &response)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", t.logger)
		return
	}

	if response.Message == "coverage" {
		successResponse = true
	}

	return
}

/*
CheckCoverageZone - For a given coverage_zone we need then to check its coverage not only by the complete coverage_zone

Returns:

	True for coverage or False for no coverage.

Args:

	country (str): Country code in ISO 3166 alpha 2 format
	coverage_zone (str):  Coverage zone
*/
func (t *CoverageApi) CheckCoverageZone(country, coverageZone string) (successResponse bool, errorResponse *responses.ErrorResponse) {
	t.logger.Info("Calling CoverageApi.CheckCoverageZone method")

	if len(country) == 0 {
		errorResponse = responses.NewError("country cannot be empty", "001", "country", t.logger)
		return
	}

	if len(coverageZone) == 0 {
		errorResponse = responses.NewError("coverageZone cannot be empty", "001", "coverageZone", t.logger)
		return
	}

	country = t.checkCountry(country)
	url := fmt.Sprintf("%s%s?country=%s&coverage_zone=%s", t.domain, t.path.Coverage, country, coverageZone)

	resp, err := t.client.Get(url, true, t.audience)
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		if resp.StatusCode == 404 {
			return
		} else {
			errorResponse = responses.NewError(fmt.Sprintf("Received error: %s", resp.Status), strconv.Itoa(resp.StatusCode), "", t.logger)
			return
		}
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while reading response body: %s", err), "", "", t.logger)
		return
	}

	var response *responses.CheckCoverageZoneResponse
	err = json.Unmarshal(body, &response)
	if err != nil {
		errorResponse = responses.NewError(fmt.Sprintf("Error while unmarshalling response body: %s", err), "", "", t.logger)
		return
	}

	if response.Message == "coverage" {
		successResponse = true
	}

	return
}

func (t *CoverageApi) checkCountry(country string) string {
	return strings.ToUpper(country)
}
